File: pset1/credit/credit.c
#include <cs50.h>
#include <stdbool.h>
#include <stdio.h>

int main(void)
{
    long cardNumber;
    do
    {
        cardNumber = get_long("Number: ");
    }
    while (cardNumber < 0);

    long tempCardNumber = cardNumber;
    int sum_of_doubled_digits = 0;
    int sum_of_other_digits = 0;
    int count = 0;
    bool is_second_digit = false;

    while (tempCardNumber > 0)
    {
        int digit = tempCardNumber % 10;

        if (is_second_digit)
        {
            int product = digit * 2;
            sum_of_doubled_digits += (product / 10) + (product % 10);
        }
        else
        {
            sum_of_other_digits += digit;
        }

        tempCardNumber /= 10;
        is_second_digit = !is_second_digit;
        count++;
    }

    int total_sum = sum_of_doubled_digits + sum_of_other_digits;
    bool luhn_valid = (total_sum % 10 == 0);

    if (luhn_valid)
    {
        long first_two_digits_calc = cardNumber;
        while (first_two_digits_calc >= 100)
        {
            first_two_digits_calc /= 10;
        }

        long first_digit_calc = cardNumber;
        while (first_digit_calc >= 10)
        {
            first_digit_calc /= 10;
        }

        if ((first_two_digits_calc == 34 || first_two_digits_calc == 37) && count == 15)
        {
            printf("AMEX\n");
        }
        else if ((first_two_digits_calc >= 51 && first_two_digits_calc <= 55) && count == 16)
        {
            printf("MASTERCARD\n");
        }
        else if (first_digit_calc == 4 && (count == 13 || count == 16))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: pset6/sentimental-credit/credit.py
import cs50

while True:
    card_number_str = cs50.get_string("Number: ")
    if card_number_str.isdigit():
        break

length = len(card_number_str)
sum1 = 0
sum2 = 0

# Luhn's Algorithm - Sum of digits multiplied by 2 (from second-to-last)
for i in range(length - 2, -1, -2):
    digit = int(card_number_str[i])
    product = digit * 2
    sum1 += product // 10 + product % 10

# Luhn's Algorithm - Sum of digits not multiplied by 2 (from last)
for i in range(length - 1, -1, -2):
    digit = int(card_number_str[i])
    sum2 += digit

total_sum = sum1 + sum2

if total_sum % 10 == 0:
    if length == 15 and (card_number_str.startswith("34") or card_number_str.startswith("37")):
        print("AMEX")
    elif length == 16 and card_number_str[0:2] in ["51", "52", "53", "54", "55"]:
        print("MASTERCARD")
    elif (length == 13 or length == 16) and card_number_str.startswith("4"):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")
